File: mobile/lib/features/products/data/models/product_model.dart
class ProductModel {
  final String id;
  final String name;
  final String sku;
  final String? barcode;
  final String? categoryId;
  final String categoryName;
  final String? description;
  final double price;
  final double costPrice;
  final int stockQuantity;
  final int minStockLevel;
  final List<String> sizes;
  final List<String> colors;
  final String? imageUrl;
  final String status;

  ProductModel({
    required this.id,
    required this.name,
    required this.sku,
    this.barcode,
    this.categoryId,
    required this.categoryName,
    this.description,
    required this.price,
    required this.costPrice,
    required this.stockQuantity,
    required this.minStockLevel,
    required this.sizes,
    required this.colors,
    this.imageUrl,
    required this.status,
  });

  factory ProductModel.fromJson(Map<String, dynamic> json) {
    return ProductModel(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      sku: json['sku'] ?? '',
      barcode: json['barcode'],
      categoryId: json['categoryId'],
      categoryName: json['categoryName'] ?? 'General',
      description: json['description'],
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      costPrice: (json['costPrice'] as num?)?.toDouble() ?? 0.0,
      stockQuantity: (json['stockQuantity'] as num?)?.toInt() ?? 0,
      minStockLevel: (json['minStockLevel'] as num?)?.toInt() ?? 5,
      sizes: (json['sizes'] as List<dynamic>?)?.map((e) => e.toString()).toList() ?? [],
      colors: (json['colors'] as List<dynamic>?)?.map((e) => e.toString()).toList() ?? [],
      imageUrl: json['imageUrl'],
      status: json['status'] ?? 'AVAILABLE',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'sku': sku,
      if (barcode != null) 'barcode': barcode,
      if (categoryId != null) 'categoryId': categoryId,
      'categoryName': categoryName,
      if (description != null) 'description': description,
      'price': price,
      'costPrice': costPrice,
      'stockQuantity': stockQuantity,
      'minStockLevel': minStockLevel,
      'sizes': sizes,
      'colors': colors,
      if (imageUrl != null) 'imageUrl': imageUrl,
    };
  }
}

File: server/src/modules/products/product.model.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IProduct extends Document {
  name: string;
  sku: string;
  barcode?: string;
  categoryId?: mongoose.Types.ObjectId;
  categoryName: string;
  description?: string;
  price: number;
  costPrice?: number;
  stockQuantity: number;
  minStockLevel: number;
  sizes: string[];
  colors: string[];
  imageUrl?: string;
  status: 'AVAILABLE' | 'LOW_STOCK' | 'OUT_OF_STOCK';
  createdAt: Date;
  updatedAt: Date;
}

const ProductSchema: Schema = new Schema(
  {
    name: { type: String, required: true, trim: true },
    sku: { type: String, required: true, unique: true, uppercase: true, trim: true },
    barcode: { type: String, trim: true },
    categoryId: { type: Schema.Types.ObjectId, ref: 'Category' },
    categoryName: { type: String, required: true, default: 'General' },
    description: { type: String, trim: true },
    price: { type: Number, required: true, min: 0 },
    costPrice: { type: Number, default: 0, min: 0 },
    stockQuantity: { type: Number, required: true, default: 0, min: 0 },
    minStockLevel: { type: Number, default: 5, min: 0 },
    sizes: [{ type: String, trim: true }],
    colors: [{ type: String, trim: true }],
    imageUrl: { type: String, trim: true },
    status: {
      type: String,
      enum: ['AVAILABLE', 'LOW_STOCK', 'OUT_OF_STOCK'],
      default: 'AVAILABLE',
    },
  },
  { timestamps: true }
);

// A scanner code must resolve to exactly one product; sparse so products
// without a barcode (undefined) don't collide.
ProductSchema.index({ barcode: 1 }, { unique: true, sparse: true });

// Pre-save middleware to calculate status automatically based on stock
ProductSchema.pre<IProduct>('save', function (next) {
  if (this.stockQuantity <= 0) {
    this.status = 'OUT_OF_STOCK';
  } else if (this.stockQuantity <= this.minStockLevel) {
    this.status = 'LOW_STOCK';
  } else {
    this.status = 'AVAILABLE';
  }
  next();
});

export const ProductModel = mongoose.model<IProduct>('Product', ProductSchema);
